### PR TITLE
[#148768631] Update to feature/buckets, using indented jinja and refactoring of complexity

### PR DESCRIPTION
**Pull request by christian-casecommons**
https://www.pivotaltracker.com/story/show/148768631

A redux of https://github.com/Casecommons/aws-cloudformation/pull/3 with the intention of refactoring for indentation and simplicity

### CODE REVIEW

**Comment by christian-casecommons** [comment]
The docs for doing so can be found on the frontend-aws pull request:
https://github.com/Casecommons/frontend-aws/pull/1

In a nutshell, here is how I would create a single bucket and associate it to a single policy:
./test/group_vars/vars.yml
```
Sts.Role: arn:aws:iam::$accountid:role/admin

config_s3_buckets:
  react-template-app.callowayart.com:
    AccessControl: PublicRead
    DeletionPolicy: Delete
    WebsiteConfiguration:
      IndexDocument: index.html
      ErrorDocument: index.html
    Policy:
      PublicReadForGetBucketObjects:
          Effect: Allow
          Principal: "*"
          Action:
            - s3:GetObject
          Resource: arn:aws:s3:::react-template-app.callowayart.com/*
```


To orchestrate, we execute the playbook as expected
```
AWS_PROFILE=$profile ansible-playbook ./site.yml -eenv=test -vv
```

As far as the use of config_s3* literals go, I followed the pattern for interpolation against cloudformation templates used throughout the accelerator stack. I've left Stack.Inputs specifically for injecting arguments against cloudformation template parameters. 
```
config_s3_buckets:
...
```

I've used jinja interpolation, as opposed to cf parameters, against this templates because it allows for defining multiple buckets and their policies, and cloudformation does not allow for iterables in any form 
https://stackoverflow.com/questions/21978641/in-aws-cloudformation-is-it-possible-to-create-a-number-of-ec2-instances-with-th

**Comment by mixja** [comment]
A few points:

- This is a good use case for Jinja (iteration is difficult with CloudFormation)
- Whenever I use Jinja to create a configuration "DSL" I typically try to reduce the amount of boilerplate required - i.e Jinja needs to be doing some work for me, it needs to be opinionated.  This is particularly important for enforcing a particular approach or pattern as to how you use the resource.
- The idiomatic approach with the new "dot/path" notation is to use `Config.<Level>...<Key>` 
- CloudFormation resource names can only be alphanumeric - and I follow Pascal naming  (e.g. react-template-app.callowayart.com should become ReactTemplateAppCallowayArtBucket or similar)

I would say with S3 buckets it's very important to understand when you should use resource policies vs IAM policies for permissions (especially in light of recent S3 exposures).  My guidance is that resource policies (i.e. bucket policies) should only be used for the following use cases:

- Granting public read only access to all objects in the bucket (common for static web assets)
- Granting read only or read/write access to other accounts (common for cross-account replication or accessing build artifacts in build accounts from other accounts)

In other words, you would never use bucket policies for granting a specific user or role access to the bucket (that is the domain of IAM policies).  Based upon the above, we could simplify the configuration DSL to a point that it only allows these two specific use cases and does not allow deviation outside of the use cases.

Of course if you want more granular control - e.g. granting permissions to only specific objects or paths in the bucket, then your approach gives this flexibility but the point is to ask yourself if this is required to ensure best practice use.

So in summary, here is how my version of the configuration DSL might look, assuming the two access use cases are valid and you con't require object-level permissions:

```
Sts.Role: arn:aws:iam::$accountid:role/admin

Config.Buckets:
  ReactTemplateAppCallowayArt:
    AccessControl: PublicRead  # This might be the only canned ACL you would allow for allowing public access to all S3 bucket objects
    DeletionPolicy: Delete
    WebsiteConfiguration:
      IndexDocument: index.html
      ErrorDocument: index.html
   ReadOnlyAccounts:
      - <account-id>
   ReadWriteAccounts:
     - <account-id>
```

**Comment by christian-casecommons** [comment]
Thanks for the feedback @mixja - ill apply your recommendations and resubmit

**Comment by christian-casecommons** [comment]
"CloudFormation resource names can only be alphanumeric - and I follow Pascal naming (e.g. react-template-app.callowayart.com should become ReactTemplateAppCallowayArtBucket or similar)"

I had one issue though: resource names are being regex stripped of illegal characters, as has been done in the past. 
https://github.com/Casecommons/aws-cloudformation/blob/master/templates/security.yml.j2#L34

Given that, I like the idea of the key being the bucket name, and in this case, the fqdn of the site being served. To me, the resource logical-id holds little value, so I don't see an issue in stripping it at "runtime". 

Let me know your thoughts @mixja

**Comment by christian-casecommons** [comment]
@mixja 

I opted to go with a combination of your recommendation (for resource names) and allow for explicitly passing a bucket name, as seen here:
https://github.com/christian-casecommons/aws-cloudformation/blob/0187bfab066399fdb1ae4f0971cfc817a7ab43d9/templates/s3.yml.j2#L7-L10

Would be specified in vars as:
```
Config.Buckets:
  ReactTemplateApp:
    BucketName: react-template-app.{{ env }}.casebookplatform.org
```

File: templates/s3.yml.j2 [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+          Effect: Allow
+          Principal:
+            AWS:
+              {% for id in config.ReadOnlyAccounts|default([]) -%}
+              - arn:aws:iam::{{ id }}:root
+              {%- endfor %}
+
+          Resource: arn:aws:s3:::{{ name }}/*
+          Action:
+            - s3:GetObject
```
Comment (pemacc) [inline templates/s3.yml.j2]: What about listing Objects. It may not even be able to see this objects.

Comment (pemacc) [inline templates/s3.yml.j2]: Test it out.

Comment (christian-casecommons) [inline templates/s3.yml.j2]:
```
ccuser@morningside:~/workspace/frontend-aws[feature/cloudfront]$ AWS_PROFILE=christian.calloway.casebook_platform aws s3api get-bucket-policy --bucket testbucket.dev.casebookplatform.org
{
    "Policy": "{\"Version\":\"2008-10-17\",\"Statement\":[{\"Sid\":\"ReactTemplateAppGetReadOnlyAccountsBucketPolicyStatement\",\"Effect\":\"Allow\",\"Principal\":{\"AWS\":\"arn:aws:iam::009823428922:root\"},\"Action\":\"s3:GetObject\",\"Resource\":\"arn:aws:s3:::testbucket.dev.casebookplatform.org/*\"},{\"Sid\":\"ReactTemplateAppListReadOnlyAccountsBucketPolicyStatement\",\"Effect\":\"Allow\",\"Principal\":{\"AWS\":\"arn:aws:iam::009823428922:root\"},\"Action\":[\"s3:GetBucketLocation\",\"s3:ListBucket\"],\"Resource\":\"arn:aws:s3:::testbucket.dev.casebookplatform.org\"}]}"
}
ccuser@morningside:~/workspace/frontend-aws[feature/cloudfront]$ AWS_PROFILE=christian.calloway.casebook_platform aws s3 cp /tmp/helloworld s3://testbucket.dev.casebookplatform.org
upload: ../../../../tmp/helloworld to s3://testbucket.dev.casebookplatform.org/helloworld
ccuser@morningside:~/workspace/frontend-aws[feature/cloudfront]$ AWS_PROFILE=testbucket aws sts get-caller-identity
{
    "Account": "009823428922",
    "UserId": "AIDAJLQ4HJ2SR3Y7AYON4",
    "Arn": "arn:aws:iam::009823428922:user/testbucket"
}
ccuser@morningside:~/workspace/frontend-aws[feature/cloudfront]$ AWS_PROFILE=testbucket aws s3 ls s3://testbucket.dev.casebookplatform.org2017-08-02 17:54:49         11 helloworld
ccuser@morningside:~/workspace/frontend-aws[feature/cloudfront]$ AWS_PROFILE=testbucket aws s3 cp s3://testbucket.dev.casebookplatform.org/helloworld /tmp
download: s3://testbucket.dev.casebookplatform.org/helloworld to ../../../../tmp/helloworld
```

Comment (christian-casecommons) [inline templates/s3.yml.j2]: Above is messy, but it works

File: templates/s3.yml.j2 [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+              {% for id in config.ReadOnlyAccounts|default([]) %}
+              - arn:aws:iam::{{ id }}:root
+              {% endfor %}
+          Resource: arn:aws:s3:::{{ name }}
+          Action:
+            - s3:GetBucketLocation
+            - s3:ListBucket
+  {% endif %}
+
+  {{ resource }}BucketLogGroup:
```
Comment (christian-casecommons) [inline templates/s3.yml.j2]: @pemacc LogGroup added; there aren't really applicable cloudwatch statistics for S3 that make sense in an alarm. 

http://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/s3-metricscollected.html

**Pull request merged**